server/memphis_handlers_tags.go

```diff
@@ -154,165 +148,254 @@ func AddTagsToEntity(tags []models.CreateTag, entity_type string, entity_id prim
+	var body models.CreateTag
 	ok := utils.Validate(c, &body, false, nil)
 	if !ok {
 		return
 	}
-	station_name, err := StationNameFromStr(body.EntityName)
+	name := strings.ToLower(body.Name)
+	exist, _, err := IsTagExist(name)
 	if err != nil {
-		serv.Errorf("CreateTags error: " + err.Error())
 		c.AbortWithStatusJSON(500, gin.H{"message": "Server error"})
 		return
 	}
-	exist, station, err := IsStationExist(station_name)
-	if err != nil {
-		serv.Errorf("CreateTags error: " + err.Error())
-		c.AbortWithStatusJSON(500, gin.H{"message": "Server error"})
+	if exist {
+		serv.Warnf("CreateNewTag error: Tag with the same name already exists")
+		c.AbortWithStatusJSON(configuration.SHOWABLE_ERROR_STATUS_CODE, gin.H{"message": "Tag with the same name already exists"})
 		return
 	}
-	if !exist {
-		serv.Warnf("Station does not exist")
-		c.AbortWithStatusJSON(configuration.SHOWABLE_ERROR_STATUS_CODE, gin.H{"message": "Could not create tags, station does not exist"})
-		return
+	var color string
+	if len(body.Color) > 0 {
+		color = body.Color
+	} else {
+		color = "101, 87, 255" // default memphis-purple color
+	}
+	var newTag models.Tag
+	stationArr := []primitive.ObjectID{}
+	schemaArr := []primitive.ObjectID{}
+	userArr := []primitive.ObjectID{}
+	newTag = models.Tag{
+		ID:       primitive.NewObjectID(),
+		Name:     name,
+		Color:    color,
+		Stations: stationArr,
+		Schemas:  schemaArr,
+		Users:    userArr,
 	}
-	err = AddTagsToEntity(body.Tags, body.EntityType, station.ID)
+
+	filter := bson.M{"name": newTag.Name}
+	update := bson.M{
+		"$setOnInsert": bson.M{
+			"_id":      newTag.ID,
+			"name":     newTag.Name,
+			"color":    newTag.Color,
+			"stations": newTag.Stations,
+			"schemas":  newTag.Schemas,
+			"users":    newTag.Users,
+		},
+	}
+	opts := options.Update().SetUpsert(true)
+	_, err = tagsCollection.UpdateOne(context.TODO(), filter, update, opts)
 	if err != nil {
-		serv.Errorf("Failed creating tag: %v", err.Error())
 		c.AbortWithStatusJSON(500, gin.H{"message": "Server error"})
 		return
 	}
-	c.IndentedJSON(200, []string{})
+
+	c.IndentedJSON(200, newTag)
 }
 
-func (th TagsHandler) RemoveTags(c *gin.Context) {
-	var body models.RemoveTagsSchema
+func (th TagsHandler) RemoveTag(c *gin.Context) {
+	var body models.RemoveTagSchema
 	ok := utils.Validate(c, &body, false, nil)
 	if !ok {
 		return
 	}
-	for _, tagName := range body.Names {
-		name := strings.ToLower(tagName)
-		exist, tag, err := IsTagExist(name)
+	name := strings.ToLower(body.Name)
+	exist, tag, err := IsTagExist(name)
 
+	if err != nil {
+		serv.Errorf("RemoveTag error: " + err.Error())
+		c.AbortWithStatusJSON(500, gin.H{"message": "Server error"})
+		return
+	}
+	if !exist {
+		c.IndentedJSON(200, []string{})
+		return
+	}
+	entity := strings.ToLower(body.EntityType)
+	switch entity {
+	case "station":
+		station_name, err := StationNameFromStr(body.EntityName)
+		if err != nil {
+			serv.Errorf("RemoveTag error: " + err.Error())
```

Comment:
when this specific function returns an error this is only a warning and a 666 error, so please change it, don't forget to change the server error as well to err.Error()